### PR TITLE
Add and use Jenkinsfile

### PR DESCRIPTION
**Pull request by stephendonner**
@m8ttyB @oremj @davehunt thoughts/r?

This can be seen passing, here: https://fx-test-jenkins.stage.mozaws.net/job/go-bouncer.pipeline/

### CODE REVIEW

File: Jenkinsfile [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+
+/** Tox environment */
+def environment = 'e2e'
+
+/** Run Tox
+ *
+ * @param environment test environment to run
+*/
+def runTox(environment) {
+  def options = env.PYTEST_ADDOPTS ?: ''
```
Comment (oremj) [inline Jenkinsfile]: The env is already passed to sh, so you don't need to set them in groovy as well.

File: Jenkinsfile [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+      deleteDir()
+      sh 'mkdir results'
+      dir('results') {
+        unstash environment
+      }
+      junit 'results/*.xml'
+      archiveArtifacts 'results/*'
+    }
+  }
+}
```
Comment (oremj) [inline Jenkinsfile]: Can we reduce this whole thing to:

```
stage('Test') {
  node {
    try {
    sh """
       tox -e e2e --color=yes
       tox -e flake8 
    """
    } catch (err) {
      currentBuild.result = 'FAILURE'
      ircNotify ...
      mail ...
    }
    junit 'results/*.xml'
    archiveArtifacts 'results/*'
  }
}
```

**Comment by stephendonner** [comment]
@oremj moved Jenkinsfile and tox.ini to the /tests/e2e/ dir; good call.

I'm not well-versed in Groovy, so I so far haven't-yet been able to figure out how to compact the Jenkinsfile script down, per your suggestions.  Help appreciated!

**Comment by oremj** [comment]
If you write a bash script with all of the steps you need to run in Jenkins, I can convert it to a pipeline script for you.

**Comment by davehunt** [comment]
> I'd rather not have Jenkinsfile, setup.cfg and tox.ini in root directory. Can we move them in to the tests subdirectory or move all the test stuff to another repo?

For what it's worth, none of these files are explicitly just for testing. Note that moving Jenkinsfile out of the root will also mean a non-default configuration in the pipeline job so that Jenkins knows where to find it. Moving the test stuff to another repo has in the past harmed the visibility and dev ownership of the tests, so I would strongly discourage that.

File: Jenkinsfile (hunk 1 above)
Comment (davehunt) [inline Jenkinsfile]: There's no guarantee this environment variable is set, so this line sets a default of an empty string.

File: Jenkinsfile (hunk 2 above)
Comment (davehunt) [inline Jenkinsfile]: We could certainly condense the Jenkinsfile, but there are issues with the proposal here. This file is based on a structure used for other projects that run multiple test environments, archive and publish HTML reports, and execute tests in parallel. I would suggest merging if this is working as expected, and then perhaps trimming it down as desired. Once we're comfortable switching to the declarative pipelines, then we'll simplify this by removing a lot of the try/catch/finally blocks. We could also reduce it even further by using a shared library for the IRC notifications as demonstrated here: https://github.com/mozmar/jenkins-pipeline/blob/master/vars/irc.groovy

**Comment by davehunt** [comment]
> If you write a bash script with all of the steps you need to run in Jenkins, I can convert it to a pipeline script for you.

The Jenkinsfile is a pipeline script already containing the bash commands needed to run in Jenkins. I'm not sure I understand what you're asking for here.

**Comment by oremj** [comment]
A bash script/pseudo code isn't absolutely needed. I can ready through the groovy instead.

**Comment by davehunt** [comment]
This will need rebasing/squashing, but I've just switched it to pipeline definition model (declarative pipeline). This is much nicer syntax, and a lot less code. We'll need to make sure that we have the pipeline definition model plugin installed on target Jenkins instances before we merge this.

**Comment by davehunt** [comment]
> I would probably try to combine most of the logic here in to one "node{}" block. The stashing and unstashing can be slow. I'd also try to simplify the try/catch/finally logic.

All done, with declarative pipelines now. So much nicer. I caught up with the Jenkins team at FOSDEM at the weekend, and heard that it reached 1.0 and was safe to start using. 😄

**Comment by davehunt** [comment]
@stephendonner I would suggest moving the requirements directory back to the root, unless @oremj objects? I think tests/e2e/requirements/tests is unnecessarily redundant.

File: tests/e2e/Jenkinsfile [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+      steps {
+        timeout(time: 5, unit: 'MINUTES') {
+          runTox(config, environment)
+        }
+      }
+    }
+  }
+  post {
+    always {
+      junit 'tests/e2e/results/*.xml'
```
Comment (bitwiseman) [inline tests/e2e/Jenkinsfile]: Suggest moving this to `success` section.

File: tests/e2e/Jenkinsfile [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+            sh "tox -c tests/e2e/tox.ini -e e2e"
+          }
+        }
+      }
+    }
+  }
+  post {
+    always {
+      junit 'tests/e2e/results/*.xml'
+      archiveArtifacts 'tests/e2e/results/*'
```
Comment (bitwiseman) [inline tests/e2e/Jenkinsfile]: Suggest you change this to: 
```
archiveArtifacts artifacts: 'tests/e2e/results/*', allowEmptyArchive: true
junit 'tests/e2e/results/*.xml'
```

Or even better, If the Lint phase doesn't produce JUnit output, you might move the `junit` step into a `post` block inside the "Test" stage.

See [JENKINS-41239](https://issues.jenkins-ci.org/browse/JENKINS-41239)

File: tests/e2e/Jenkinsfile [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+  stages {
+    stage('Lint') {
+      steps {
+        sh "tox -c tests/e2e/tox.ini -e flake8"
+      }
+    }
+    stage('Test') {
+      steps {
+        timeout(time: 5, unit: 'MINUTES') {
+          wrap([$class: 'AnsiColorBuildWrapper']) {
```
Comment (bitwiseman) [inline tests/e2e/Jenkinsfile]: In declarative, you can remove this wrap and add `ansiColor` to the `options` block:

```
options {
    ansiColor()
    timestamps()
}
```

Comment (bitwiseman) [inline tests/e2e/Jenkinsfile]: You can do the same with `timeout` if you're okay with it being for the pipeline as a whole:

options {
    ansiColor()
    timestamps()
    timeout time:5, units:'MINUTES'
}

**Comment by bitwiseman** [comment]
It is really great to see Declarative Pipeline here!

File: tests/e2e/Jenkinsfile (hunk 4 above)
Comment (davehunt) [inline tests/e2e/Jenkinsfile]: Interesting! I didn't realise that stages can have post blocks! 👍

**Comment by davehunt** [comment]
> It is really great to see Declarative Pipeline here!

Thanks so much for your suggestions @bitwiseman!

@stephendonner @oremj this is rebased/squashed and ready for final review and merge. The new declarative pipelines make it so much nicer. In the near future I'll probably switch the `ircNotification` to use a shared library. Also, it would be good to switch to using a docker agent to remove the need for Python and Tox to be on the Jenkins agents.

**Comment by bitwiseman** [comment]
All looks reasonable to me.  Great job.

**Comment by oremj** [comment]
Merging per @stephendonner request.

**Pull request merged**